xin_kai_vertical_diffusivity.jl

```diff
@@ -0,0 +1,255 @@
+    ∂z_v² = ℑyᵃᶜᵃ(i, j, k, grid, ϕ², ∂zᶜᶠᶠ, velocities.v)
+    return ∂z_u² + ∂z_v²
+end
+
+@inline function N²ᶜᶜᶠ(i, j, k, grid, buoyancy, tracers)
+    return ∂z_b(i, j, k, grid, buoyancy, tracers)
+end
+
+@inline function Riᶜᶜᶠ(i, j, k, grid, velocities, diffusivities)
+    S² = shear_squaredᶜᶜᶠ(i, j, k, grid, velocities)
+    N² = diffusivities.N²[i, j, k]
+    Ri = N² / S²
+
+    # Clip N² and avoid NaN
+    return ifelse(N² == 0, zero(grid), Ri)
+end
+
+const c = Center()
+const f = Face()
+
+@kernel function compute_N²!(diffusivities, grid, closure::FlavorOfXKVD, tracers, buoyancy)
+    i, j, k = @index(Global, NTuple)
+    @inbounds diffusivities.N²[i, j, k] = N²ᶜᶜᶠ(i, j, k, grid, buoyancy, tracers)
+end
+
+@kernel function compute_ri_number!(diffusivities, grid, closure::FlavorOfXKVD, velocities)
+    i, j, k = @index(Global, NTuple)
+    @inbounds diffusivities.Ri[i, j, k] = Riᶜᶜᶠ(i, j, k, grid, velocities, diffusivities)
+end
+
+@kernel function compute_xinkai_diffusivities!(diffusivities, grid, closure::FlavorOfXKVD,
+                                                velocities, tracers, buoyancy, tracer_bcs, clock)
+    i, j, k = @index(Global, NTuple)
+    _compute_xinkai_diffusivities!(i, j, k, diffusivities, grid, closure,
+                                   velocities, tracers, buoyancy, tracer_bcs, clock)
+end
+
+@inline function _compute_xinkai_diffusivities!(i, j, k, diffusivities, grid, closure,
+                                                velocities, tracers, buoyancy, tracer_bcs, clock)
+
+    # Ensure this works with "ensembles" of closures, in addition to ordinary single closures
+    closure_ij = getclosure(i, j, closure)
+
+    ν₀  = closure_ij.ν₀  
+    νˢʰ = closure_ij.νˢʰ
+    νᶜⁿ = closure_ij.νᶜⁿ
+    Cᵉⁿ = closure_ij.Cᵉⁿ
+    Prₜ = closure_ij.Prₜ
+    Riᶜ = closure_ij.Riᶜ
+    δRi = closure_ij.δRi
+    Q₀  = closure_ij.Q₀ 
+    δQ  = closure_ij.δQ 
+
+    Qᵇ = top_buoyancy_flux(i, j, grid, buoyancy, tracer_bcs, clock, merge(velocities, tracers))
+
+    # (Potentially) apply a horizontal filter to the Richardson number
+    Ri = ℑxyᶜᶜᵃ(i, j, k, grid, ℑxyᶠᶠᵃ, diffusivities.Ri)
+    Ri_above = ℑxyᶜᶜᵃ(i, j, k + 1, grid, ℑxyᶠᶠᵃ, diffusivities.Ri)
+    N² = ℑxyᶜᶜᵃ(i, j, k, grid, ℑxyᶠᶠᵃ, diffusivities.N²)
+    
+    # Conditions
+    convecting = Ri < 0 # applies regardless of Qᵇ
+    entraining = (Ri > 0) & (Ri_above < 0) & (Qᵇ > 0)
+
+    # Convective adjustment diffusivity
+    ν_local = ifelse(convecting, - (νᶜⁿ - νˢʰ) / 2 * tanh(Ri / δRi) + νˢʰ, clamp(Riᶜ * Ri + νˢʰ + ν₀, ν₀, νˢʰ))
+
+    # Entrainment diffusivity
+    x = Qᵇ / (N² + 1e-11)
+    ν_nonlocal = ifelse(entraining,  Cᵉⁿ * νᶜⁿ * 0.5 * (tanh((x - Q₀) / δQ) + 1), 0)
+
+    # Update by averaging in time
+    @inbounds diffusivities.κᵘ[i, j, k] = ifelse(k <= 1 || k >= grid.Nz+1, 0, ν_local + ν_nonlocal)
+    @inbounds diffusivities.κᶜ[i, j, k] = ifelse(k <= 1 || k >= grid.Nz+1, 0, (ν_local + ν_nonlocal) / Prₜ)
```

Comment:
```suggestion
    @inbounds diffusivities.κᵘ[i, j, k] = ifelse((k <= 1) | (k >= grid.Nz+1), 0, ν_local + ν_nonlocal)
    @inbounds diffusivities.κᶜ[i, j, k] = ifelse((k <= 1) | (k >= grid.Nz+1), 0, (ν_local + ν_nonlocal) / Prₜ)
```

Reply:
out of curiosity, why do we prefer bitwise OR over the short circuiting version?

Reply:
out of curiosity, why do we prefer bitwise OR over the short circuiting version?